### ISSUE TITLE
gracefull upgrades through terraform

### ISSUE THREAD

**Issue by camaeel** [siderolabs/terraform-provider-talos#140]
Currently there is no option to trigger talos upgrade through terraform provider.
If I aplly new kubernetes version through talos resources all nodes get patched simultaneously.
It would be great to have resources to upgrade both talos itself and kubernetes in a gracefull way one by one node (like `talosctl upgrade` and `taloscyl upgrade-k8s` does).

Another option could be, to have builtin upgrade controller in talos controllers.

**Comment by M4t7e** [comment]
IMHO, this is by far one of the most important functionalities missing in the Talos Terraform provider. To manage a Talos K8s cluster using only Terraform, we need a declarative and idempotent way to specify a target version. The lifecycle of a cluster should be considered more thoroughly. The Talos and Kubernetes versions also have be compatible with the additional components that may be needed, such as external CCM, CNI, CSI, etc. Talos introduces imperative change management here, which is undesirable in a Terraform environment. Therefore, I don't see any feasible way to conduct proper lifecycle management with this module.

**Comment by hegerdes** [comment]
As a workaround you can specify the `parallelism` [Terrafrom Docs](https://developer.hashicorp.com/terraform/cli/commands/apply#parallelism-n) arg and setting it to `1` default is `10`. So terraform will only do one resource at the time. When you pair this with the terraform `time` resource you can archive wait till the node is done. 

Not pretty but bast we can do right now besides from gracefully replacing nodes

**Comment by perfectra1n** [comment]
Yeah, without a way to run `talosctl upgrade` when changing Talos versions or extensions via this provider is rough - it makes all the work done on this provider for naught. It is possible to first create the cluster using this provider, but being unable to upgrade the cluster through this provider is a huge bummer.

**Comment by nwmcsween** [comment]
This could be done with different "vm instance/server/etc" blocks, the initial block would create and destroy on any update and the second block would depend on data http to connect before creating/destroying.

**Comment by divStar** [comment]
I actually am interested in this, too. I suppose one could use `local-exec`, but it becomes very ugly very quickly.
There is also some idea I saw in https://blog.stonegarden.dev/articles/2024/08/talos-proxmox-tofu/, but to be perfectly honest I don't understand how it works.

And since I am on my homelab server with somewhat-limited resources, I wouldn't mind some downtime while upgrading, but I'd need it to e.g. destroy the current nodes and create new ones, mount the disk device blocks and boot everything - just like it currently does, but with VMs with newer versions of Talos.

Isn't it enough to just destroy and re-create the nodes with newer versions? Or should we prefer to "upgrade" running nodes?

**Comment by amaol-vestas** [comment]
We need this feature in talos terraform provider, or the alternative is just a nasty bad code practice, please @smira take this need into account, thanks

**Comment by ionfury** [comment]
I have the same issue.  I implemented a nasty workaround with [local-exec](https://github.com/ionfury/homelab-modules/blob/8f08e0e11de46aab20ac8238a25d397b47c013ea/modules/talos-cluster/image.tf#L39) and a [script](https://github.com/ionfury/homelab-modules/blob/8f08e0e11de46aab20ac8238a25d397b47c013ea/modules/talos-cluster/resources/scripts/upgrade-node.sh).

**Comment by Auroch** [comment]
I'm also very interested in this function. If you're working on it, think about the --preserve option in talosctl so that you can choose not to delete EPHEMERAL (for longhorn users, for example).

**Comment by soostdijck** [comment]
This would be a good feature to have, but as a cynical person, I doubt it will happen any time soon. Why would it, Sidero would much rather have you use Omni instead 💰

**Comment by Dext00** [comment]
This is a really needed feature. That's not the only posted issue here with this topic.
@frezbo , you added this up on your board once. Why isn't it there (anymore)?
Unfortunately, I have no experience with Go, otherwise I would try to build it myself. Do you have any documentation on how to create and test contributions?

**Comment by laurivosandi** [comment]
This was not pleasant surprise

**Comment by Dext00** [comment]
I made a little workaround for my setup with proxmox vms. When a node is to be updated, it is first shut down and its VM is deleted. Subsequently, it is recreated and installed with the new version. Two key factors make this process work: firstly, the lifecycle configuration for the `resource talos_machine_configuration_apply`, and—most importantly—the VMs must be updated sequentially.
To avoid having to manually recreate every single node one after another forever, I used AI to generate a script that runs `terraform state list`, extracts the VM resources, and then executes them sequentially using `terraform apply -replace`.

```hcl
# Resource to create proxmox vms
resource "proxmox_virtual_environment_vm" "talos_vm" {
  for_each = local.all_nodes
  depends_on = [proxmox_download_file.talos_nocloud_image]

  name        = each.value.vm_name
  description = "${each.value.vm_name}: ${each.value.talos_type} node within the Kubernetes cluster - ${var.cluster_name}."
  tags        = var.vm_tags
  node_name   = each.value.host_node
  on_boot     = true
  vm_id       = each.value.vm_id
  cpu {
    cores = each.value.cpu
    type = "x86-64-v2-AES"
  }
  memory {
    dedicated = each.value.ram
  }
  agent {
    enabled = true
  }
  network_device {
    bridge = each.value.network_device
    vlan_id = var.vlan_id
  }
  boot_order = ["virtio0", "ide3"]
  disk {
    datastore_id = "local-lvm"
    interface    = "virtio0"
    size         = each.value.disk_size
  }
  cdrom {
    file_id      = resource.proxmox_download_file.talos_nocloud_image[each.value.host_node].id
  }
  operating_system {
    type = "l26" # Linux Kernel 2.6 - 5.X.
  }
  initialization {
    datastore_id = "local-lvm"
    ip_config {
      ipv4 {
        address = "${each.value.ip}/${var.subnetmask}"
        gateway = var.gateway
      }
    }
  }
}
```

```hcl
# Applies Talos machine configuration for worker nodes
# Same for control planes
resource "talos_machine_configuration_apply" "worker" {
  for_each                    = var.worker_nodes
  depends_on = [proxmox_virtual_environment_vm.talos_vm]

  client_configuration        = talos_machine_secrets.machine_secrets.client_configuration
  machine_configuration_input = data.talos_machine_configuration.worker.machine_configuration
  node                        = each.value.ip
  on_destroy = {
    graceful = false
    reboot = false
    reset = true
  }
  config_patches = [
    templatefile("${path.module}/all-nodes-config-apply.yaml.tftpl", {
      installer_image_url = data.talos_image_factory_urls.image_url.urls.installer
    })
  ]

  lifecycle {
    replace_triggered_by = [
      proxmox_virtual_environment_vm.talos_vm[each.key].id
    ]
  }
}

```

**Comment by mcanevet** [comment]
`v0.12.0-alpha.0` supports declarative Talos and Kubernetes upgrade with new `talos_machine` and `talos_cluster` resources.

**Comment by eplightning** [comment]
@mcanevet 
Is it possible to use `talos_machine` and `talos_cluster` together?

From my understanding Kubernetes version is also inside the machine configuration - given that it contains container images used for Kubernetes components, including the tag (version). So upgrading K8s also modifies the machine configurations during the process.

Given that, wouldn't those resources conflict since they both affect machine config? I can't find any logic inside `talos_machine` that would ignore the Kubernetes version changes specifically.

Or am I completely misunderstanding something?

**Comment by mcanevet** [comment]
Your understanding is correct, there is a real conflict.

When `upgrade-k8s` runs (triggered by `talos_cluster`), it calls `ApplyConfiguration` with `NO_REBOOT` mode on each node, which fully replaces the persisted machine config on disk with updated image tags for kubelet, kube-apiserver, kube-controller-manager, kube-scheduler, and kube-proxy. On the next `terraform plan`, `talos_machine` sees a hash mismatch and re-applies its original config, reverting those image tags, creating an infinite conflict loop.

The safe workflow is to upgrade in two separate applies, keeping `kubernetes_version` in sync between `talos_machine_configuration` and `talos_cluster`:

1. Update `talos_version` in `talos_machine_configuration` → `terraform apply`
2. Update `kubernetes_version` in both `talos_machine_configuration` and `talos_cluster` → `terraform apply`

In step 2, both resources target the same image tags, so regardless of execution order the end state is consistent and the hash stabilises.

That said, this upgrade path needs to be validated by an integration test and clearly documented before being recommended. It has not been verified to work in practice.

**Comment by eplightning** [comment]
I see, still the problem is in step 2 that if talos_machine_configuration goes first the upgrade will no longer be graceful (like requested in this issue).

**Comment by mcanevet** [comment]
Hi @frezbo @smira 

Before submitting a fix I'd like to make sure I correctly understood the problem and that my approach makes sense.

How I understand the issue:

When `kubernetes_version` is bumped in `talos_machine_configuration`, the generated YAML gains new image tags for five fields:
- `machine.kubelet.image`
- `cluster.apiServer.image`
- `cluster.controllerManager.image`
- `cluster.scheduler.image`
- `cluster.proxy.image`

`talos_machine` detects this as a config change and calls `ApplyConfiguration` on every managed node in parallel. This bypasses `upgrade-k8s'`s sequential, health-gated procedure (pre-pull → static pod upgrade per component → kube-proxy DaemonSet → kubelet per node → manifest sync).

Meanwhile, `talos_cluster` already calls `k8s.Upgrade()` properly when its `kubernetes_version` changes. So if both resources are used together, bumping `kubernetes_version` in `talos_machine_configuration` causes `talos_machine` to race against `talos_cluster`'s sequential upgrade, or, if `talos_cluster` isn't used at all, silently applies an incomplete upgrade (no kube-proxy DaemonSet update, no manifest re-apply, no sequential ordering).

My proposed fix:

We compute the machine configuration hash by stripping those five image fields before hashing. If the hash excluding K8s images is unchanged, `talos_machine` skips `ApplyConfiguration`, even when a simultaneous Talos OS upgrade is happening. Structural config changes (anything outside those five fields) still trigger a normal apply.

Questions

1. Did I correctly identify the five fields that `upgrade-k8s` owns? Are there others we should exclude?
2. Is hash normalization (strip → hash) the right layer to enforce this, or would you prefer a different approach?
3. Any concerns about the YAML round-trip stability of stripping fields across Talos versions?

Happy to share the draft PR if that helps give context.

**Comment by smira** [comment]
I don't know what Terraform is capable of, but certainly all of these are workarounds for using wrong language (Terraform) for orchestration of complex in-place updates. 

Omni supports rolling out Kubernetes upgrades properly, and I think a better path for Terraform-heavy users is to introduce a Terraform provider for Omni, rather than trying to replicate things which are already implemented properly.

**Comment by mcanevet** [comment]
Agreed that Terraform isn't the right orchestration layer for complex in-place upgrades, and an Omni provider would be a better long-term path for teams that need that.

This proposal is narrower than that, though: it's not trying to replicate `upgrade-k8s` inside Terraform. It's about preventing `talos_machine` from interfering with `upgrade-k8s`. When `kubernetes_version` changes in `talos_machine_configuration`, the five K8s component image fields change in the generated config. Without this, `talos_machine` sees a config diff and re-applies those fields across all nodes in parallel — bypassing the sequential, health-gated procedure that `upgrade-k8s` provides. The proposal just makes `talos_machine` ignore those five fields when computing whether a re-apply is needed, so `upgrade-k8s` remains the sole owner of them.

In other words: it's about `talos_machine` knowing what it doesn't own, not about it doing more.

**Comment by frezbo** [comment]
This isn;t going to work long term, Talos 1.14 is going to deprecate those and move to multi doc for each kubernetes component config, so TF provider would have a play a game of catch up to talos always, other option is for provider to trust machineconfig from the machine as truth when refreshing

**Comment by smira** [comment]
In theory just bumping kubernetes_version after the upgrade show make config diff empty, Talos machinery will correctly patch versions back

**Comment by mcanevet** [comment]
Here is my current proposal: https://github.com/siderolabs/terraform-provider-talos/pull/350

**Comment by mcanevet** [comment]
Thanks @smira, you're right that keeping `kubernetes_version` in sync after `upgrade-k8s` makes the config diff naturally empty in the sequential workflow (upgrade cluster first, then update `talos_machine_configuration`). I'll document that workflow clearly so users know the safe sequencing. The fix in #350 is mainly a safety net for the common case where both are bumped in the same apply and I'm not sure we should do it.

**Comment by mcanevet** [comment]
https://github.com/siderolabs/terraform-provider-talos/pull/353

**Comment by ludwighansson** [comment]
This is how I ended up doing upgrades when testing the new alpha release with `talos_cluster` and `talos_machine`. During the first apply I would run with `manage_kubernetes_upgrades = false`, which would disable the `talos_cluster` resource and skip the control node resource's `depends_on = [talos_cluster.this]` as count would be zero, allowing it to continue. To bootstrap the cluster I then had to rely on the old `talos_machine_boostrap` resource. 

After the cluster had been bootstrapped and was healthy, I could then set `manage_kubernetes_upgrades = true` to allow the provider to perform Kubernetes upgrades. Next time I would update `kubernetes_version` and apply the changes, both `talos_machine_configuration` and `talos_cluster` would specify the same terraform version, which ends up in an empty diff between the terraform generated and running machine config read from the API after `talos_cluster`applied it's changes (assuming no other changes were made of course).

I guess this wasn't the intended way of combining these resources, but I value an easier upgrade flow more than bootstrapping. A bootstrap I will do once per cluster, but will have reoccurring Kubernetes upgrades for multiple clusters every now and then. So this is a trade-off I'm willing to accept.

Here is a simplified snippet of what my Terraform configuration looks like:

```hcl
resource "talos_cluster" "this" {
  count = var.manage_kubernetes_upgrades ? 1 : 0

  node                 = var.bootstrap_node
  client_configuration = talos_machine_secrets.cluster.client_configuration
  kubernetes_version   = var.kubernetes_version
  control_plane_nodes  = var.controlplane_nodes
}

resource "talos_machine" "control01" {
  client_configuration  = talos_machine_secrets.cluster.client_configuration
  machine_configuration = data.talos_machine_configuration.control01.machine_configuration
  node                  = "10.200.0.20"
  image                 = "${var.talos_image_repository}/${talos_image_factory_schematic.this.id}:${var.talos_version}"
  
  depends_on            = [talos_cluster.this]
}

resource "talos_machine_bootstrap" "this" {
  client_configuration = talos_machine_secrets.cluster.client_configuration
  node                 = var.bootstrap_node

  depends_on = [talos_machine.talos_control01]
}
```

So I guess it can be built in two ways, either you make two applies on every upgrade as the new documentation suggests in https://github.com/siderolabs/terraform-provider-talos/pull/353, or you enable the `talos_cluster resource` after the cluster has been bootstrapped and configures depends_on to `talos_machine` resources. Then you only need one commit/apply on Kubernetes upgrades.

**Comment by mcanevet** [comment]
I'll close this issue because the documentation now clearly explains how to upgrade.

**Comment by mcanevet** [comment]
https://github.com/siderolabs/terraform-provider-talos/pull/370

**Comment by supersidr** [comment]
@mcanevet Hi,

Could you please tell if your pull request (code) can update Talos version?

**Comment by mcanevet** [comment]
Yes, it can upgrade Talos and Kubernetes

**Comment by supersidr** [comment]
@mcanevet I read docs many times but didn't understand how upgrade Talos. Could you please tell how can I upgrade Talos version?

**Comment by mcanevet** [comment]
Just specify the version to use in the `talos_machine` resource. If there's a drift, terraform will reconcile.

**Comment by supersidr** [comment]
@mcanevet So, if I update the image field in talos_machine to ghcr.io/siderolabs/installer:v1.10.0, will Terraform automatically trigger a Talos upgrade?

**Comment by mcanevet** [comment]
Yes

**Comment by nerdlich** [comment]
I'm confused on how to use the new resources 'talos_machine' and 'talos_cluster'. Probably not helped by the fact that I'm new to terraform. Do I need to define those resources in addition to the 'talos_machine_configuration_apply' I have now?

**Comment by mcanevet** [comment]
No, you should not define `talos_machine_configuration_apply` anymore.
Actually this resource may be deprecated soon.
